### PR TITLE
Add X86_64Darwin target

### PR DESCRIPTION
**Pull request by refusal-type-bamboo**
X86_64Darwin.v3 added. 
Changes in MachO.v3 for 64bit.

### CODE REVIEW

File: aeneas/src/x86-64/X86_64Darwin.v3 [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+		asm.push_i(signo);		// signal number
+		asm.push_i(0);			// "dummy" value
+		asm.movd_rm_i(X86Regs.EAX, 46); // sigaction
+		asm.intK(0x80);
+		asm.add.rm_i(X86Regs.ESP, 32);	// pop params off stack
+	}
+}
+// a mach-O thread state command for x86
+class MachO_X86_64ThreadState extends MachO_LoadCmd {
+	var eip: int;
```
Comment (titzer) [inline aeneas/src/x86-64/X86_64Darwin.v3:256]: Can you change this to type `u64` (or `long`) and change the name to `rip`?

File: aeneas/src/x86-64/X86_64Darwin.v3 [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+		var s = MachO_SegmentLoad.new(name);
+		s.initprot = prot;
+		s.maxprot = prot;
+		header.addCmd(s);
+		return s;
+	}
+	}
+
+// Darwin-specific backend code generation.
+class X86_64DarwinBackend extends X86Backend {
```
Comment (titzer) [inline aeneas/src/x86-64/X86_64Darwin.v3:171]: This should extend `X86_64Backend` and the methods below can be copied/adapted from X86_64LinuxBackend

File: aeneas/src/x86-64/X86_64Darwin.v3 [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+	def computeFieldOffsets(prog: Program, b: Vector<IrField>, start: int) {
+		getRuntime(prog).mach.computeFieldOffsets(prog, b, start);
+	}
+	def addRoots(compiler: Compiler, prog: Program) {
+		getRuntime(prog).addRoots();
+	}
+	private def getRuntime(prog: Program) -> MachRuntime {
+		return MachProgram.!(prog.tprog).runtime;
+	}
+	def emitDBG(compiler: Compiler, prog: Program) {
```
Comment (titzer) [inline aeneas/src/x86-64/X86_64Darwin.v3:42]: You can remove this method now.

File: aeneas/src/x86-64/X86_64Darwin.v3 [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+component X86_64DarwinTargets {
+	new() {
+		Aeneas.registerTarget(X86_64DarwinTarget.new("x86-64-darwin-test", true));
+		Aeneas.registerTarget(X86_64DarwinTarget.new("x86-64-darwin", false));
+	}
+}
+
+def SPACE = AddressSpace.new("mem", false, 64, 8,
+	Alignment.new(4096), Alignment.new(8));
+// Darwin target for x86 architecture. Generates an Mach-O binary directly.
```
Comment (titzer) [inline aeneas/src/x86-64/X86_64Darwin.v3:13]: Can you update the comment to "x86-64"?

File: aeneas/src/x86-64/X86_64Darwin.v3 [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+		compiler.makeExecutable(file);
+	}
+	def emit(compiler: Compiler, prog: Program) {
+		var rt = getRuntime(prog), mach = rt.mach;
+		mach.build(compiler, MachLoweringConfig.new());
+
+		var header = MachO_Header.new(true);
+		var pageAlign = rt.codeRegion.space.pageAlign;
+		header.cputype = MachO.CPU_TYPE_X86_64;
+		header.cpusubtype = 3;
```
Comment (titzer) [inline aeneas/src/x86-64/X86_64Darwin.v3:83]: Can you add a constant to `MachO` for this?

File: aeneas/src/x86-64/X86_64Darwin.v3 [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+		w.skipPage();
+		// generate code and runtime tables into buffer
+		// encodeCode(w, cs, ts);
+		var backend = X86_64DarwinBackend.new(compiler, prog, mach, w, test);
+		ts.eip = w.endAddr();
+		cs.vmaddr = w.endPageAddr();
+		rt.recordCodeStart(ts.eip);
+		backend.genAllCode();
+		var size = w.end();
+		if (size < pageAlign.size) {
```
Comment (titzer) [inline aeneas/src/x86-64/X86_64Darwin.v3:113]: I think you can remove this `if` for now; we'll add this back if it is an issue.

File: aeneas/src/x86-64/X86_64Darwin.v3 [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+		compiler.makeExecutable(file);
+	}
+	def newSegmentLoad(header: MachO_Header, name: string, prot: int) ->  MachO_SegmentLoad {
+		var s = MachO_SegmentLoad.new(name);
+		s.initprot = prot;
+		s.maxprot = prot;
+		header.addCmd(s);
+		return s;
+	}
+	}
```
Comment (titzer) [inline aeneas/src/x86-64/X86_64Darwin.v3:168]: Indentation of this curly brace is off.

File: aeneas/src/x86-64/X86_64Darwin.v3 [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+}
+// a mach-O thread state command for x86
+class MachO_X86_64ThreadState extends MachO_LoadCmd {
+	var eip: int;
+	def encode(is64: bool, w: DataWriter) {
+		Terminal.put2("ThreadState command with size %d, flavor 6: starts at %d", size(true), w.pos);
+		w.put_b32(0x5); // unix thread
+		w.put_b32(21 * 8 + 4 * 4);  // 80 bytes for x86
+		w.put_b32(4);   // X86-64 thread state flavor
```
Comment (titzer) [inline aeneas/src/x86-64/X86_64Darwin.v3:261]: Can you add a constant to `MachO` for this?

File: aeneas/src/x86-64/X86_64Darwin.v3 [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+		w.put_b32(42);  // number of words
+		w.zeroN(21 * 8);
+		var current_pos = w.pos;
+		w.at(w.pos - 5 * 8);
+		w.put_b64(eip);
+		w.at(current_pos);
+		Terminal.put1(", ends at position %d\n", w.pos);
```
Comment (titzer) [inline aeneas/src/x86-64/X86_64Darwin.v3:268]: You can remove the `Terminal` output.

File: aeneas/src/exe/MachO.v3 [hunk 10]
```diff
@@ -91,27 +91,40 @@ class MachO_SegmentLoad(segname: string) extends MachO_LoadCmd {
...
 		w.put_b32(nsects);
 		w.put_b32(flags);
 		for (l = sectList; l != null; l = l.tail) {
 			l.head.encode(is64, w, this);
 		}
+		var diff = w.pos - start;
+		//write the size of the command and go back
+		w.at(sizePos).put_b32(diff);
+		w.at(start + diff);
+		Terminal.put3("Segment load command with size %d: starts at position %d, ends at position %d\n", diff, start, w.pos);
```
Comment (titzer) [inline aeneas/src/exe/MachO.v3:122]: You can remove the `Terminal` output, here and elsewhere.

**Review by titzer — COMMENTED** [review]
Looking good, PTAL at suggestions.

**Comment by refusal-type-bamboo** [comment]
Issues mentioned are fixed. Please let me know if there are further problems. Thanks for your help!

**Comment by titzer** [comment]
Can you push your changes to your GitHub branch? I only see the initial commits.

**Comment by refusal-type-bamboo** [comment]
Modified version pushed. Sorry for that!

**Review by titzer — APPROVED** [review]
lgtm.

Can you do one final thing? Run "aeneas bump" to increment the version number, and then I can merge this PR.

**Pull request merged**